### DISCUSSION TITLE
Advice for seeder plugin

### DISCUSSION THREAD

**Discussion by ofek** [pypa/virtualenv#2245] [Q&A] [answered]
I've read the docs & source code of the built-in seeders but I can't quite figure out what to do.

Essentially, all I need is to install a certain package/wheel when each virtual environment is created.

cc @gaborbernat do you have any recommendations?

**Comment by ofek** [comment]
Bump 🙂 

Does anyone have an example of this?

**Comment by gaborbernat** [comment]
Ah forgot about this, can you ping me again on Monday? 😁And I'll have a look and path forward.

**Comment by ofek** [comment]
Ping!

**Comment by gaborbernat** [comment, marked as answer]
> cc @gaborbernat do you have any recommendations?

So my original idea was that in this case, you could extend your code from an existing seeder, and then you can do it at the end of the run method:

```python
from virtualenv.seed.embed.via_app_data.via_app_data import FromAppData

class CustomSeeder(FromAppData):
    def run(self, creator):
        super().run()
        subprocess.run([self.exe, '-m', 'pip', 'install', 'magic'])
```

```ini
# setup.cfg

virtualenv.seed =
  custom = virtualenv_plugin.CustomSeeder
```

One thing that's not really solved, though, is how you enable this seeder. The user would have to opt-in to use it explicitly via config/CLI args. Of course, you can always use monkey patch the default here https://github.com/pypa/virtualenv/blob/de69dfadf6d8f622a82f269b7d43832997c86cb1/src/virtualenv/run/plugin/seeders.py#L28 ... but there's no official way at the moment to change the default seeder without explicit opt-in.

**Reply by studioj** [reply]
is there somewhere an example on how to use virtualenv plugins? I'll gladly use a custom seeder to auto install (in a speedy way) packages in a CI system whenever the virtualenv is created. Just as @acmpo6ou stated for rich.
But I dont see how to get started. (maybe I can extend the docs whenever I find a good example)
starting from here https://virtualenv.pypa.io/en/20.16.5/extend.html#extend-functionality
... does this mean it is enabled by default whenever the plugin is installed?

**Comment by ofek** [comment]
Thanks! If I ship the wheel, is there a way to add it to the list of embedded wheels so it automatically gets installed? I tried `--extra-search-dir` but that did not work.

**Comment by gaborbernat** [comment]
You can extend the list of packages to install under https://github.com/pypa/virtualenv/blob/main/src/virtualenv/seed/embed/base_embed.py#L38. You might need to patch the embed file list under https://github.com/pypa/virtualenv/blob/main/src/virtualenv/seed/wheels/embed/__init__.py#L7 to get picked up. Tag me as a reviewer if you can, so can double-check how you do it looks good.

**Comment by acmpo6ou** [comment]
Hey guys!

I also need this, I need a certain package (`rich`) to be installed in every venv I create with virtualenv. But I need it to be fast, so that when I create a venv it is created fast and it contains `pip`, `wheel`, `setuptools` and `rich`. It probably has to use the magic `app-data` and not `pip` when installing `rich`.

It seems that you guys are talking exactly about this, aren't you? Do you know how to configure virtualenv, so that it does what I need?

Thank you!

**Comment by gaborbernat** [comment]
I don't think that plugin has been created, so you'd likely need to do so 👍

**Comment by gaborbernat** [comment]
This discussion has an accepted answer with a detailed example of how to create a custom seeder plugin by extending `FromAppData`. If there's interest in making it easier to configure a default custom seeder without explicit opt-in, please open a new feature request issue with specific use cases. Closing as answered.
